pika/adapters/blocking_connection.py

```diff
@@ -699,8 +699,10 @@ def process_data_events(self, time_limit=0):
             until I/O produces actionalable events. Defaults to 0 for backward
             compatibility. This parameter is NEW in pika 0.10.0.
         """
-        common_terminator = lambda: bool(
-            self._channels_pending_dispatch or self._ready_events)
+        with self._acquire_event_dispatch() as dispatch_acquired:
+            # Check if we can actually process pending events
+            common_terminator = lambda: bool(dispatch_acquired and
+                (self._channels_pending_dispatch or self._ready_events))
 
         if time_limit is None:
```

Comment:
Once we get to this line, the `yield` in `_acquire_event_dispatch()` has been exited and we no longer hold the resource. The `if` / `else` statements *must* be indented to be part of the `with` block.